src/erl_cache_server.erl

```diff
@@ -69,24 +70,48 @@ start_link(Name) ->
 get(Name, Key, WaitForRefresh) ->
     Now = now_ms(),
     case ets:lookup(get_table_name(Name), Key) of
-        [#cache_entry{validity=Validity, value=Value}] when Now < Validity ->
-            gen_server:cast(Name, {increase_stat, hit}),
-            {ok, Value};
-        [#cache_entry{evict=Evict, value=Value, refresh_callback=undefined}] when Now < Evict ->
-            gen_server:cast(Name, {increase_stat, overdue}),
-            {ok, Value};
-        [#cache_entry{evict=Evict, refresh_callback=Cb}=Entry] when Now < Evict, Cb /=undefined ->
-            ?DEBUG("Refreshing overdue key ~p", [Key]),
-            gen_server:cast(Name, {increase_stat, overdue}),
-            {ok, NewVal} = refresh(Name, Entry, WaitForRefresh),
-            {ok, NewVal};
-        [#cache_entry{value=_ExpiredValue}] ->
-            {error, not_found};
+        [Entry] ->
+            send_stat(Name, Now, Entry),
+            case get_valid_value(Name, WaitForRefresh, Now, Entry) of
+                {true, Value} -> {ok, Value};
+                false -> {error, not_found}
+            end;
         [] ->
             gen_server:cast(Name, {increase_stat, miss}),
             {error, not_found}
     end.
 
+send_stat(Name, Now, #cache_entry{validity=Validity}) when Now < Validity ->
+    gen_server:cast(Name, {increase_stat, hit});
+send_stat(Name, Now, #cache_entry{evict=Evict}) when Now < Evict ->
+    gen_server:cast(Name, {increase_stat, overdue});
+send_stat(_Name, _Now,  _) ->
+    ok.
+
+get_valid_value(_Name, _IsSync, Now, #cache_entry{validity=Validity} = E) when Now < Validity ->
+    {true, E#cache_entry.value};
+get_valid_value(Name, IsSync, Now, #cache_entry{evict=Evict} = E) when Now < Evict ->
+    refresh(Name, E, IsSync);
+get_valid_value(_Name, _IsSync, _Now, #cache_entry{value=_ExpiredValue}) ->
+    false.
+
+-spec match(erl_cache:name(), erl_cache:key(), erl_cache:wait_for_refresh()) ->
+    {ok, erl_cache:value()} | {error, not_found}.
+match(Name, Key, WaitForRefresh) ->
+    Now = now_ms(),
+    case ets:match_object(get_table_name(Name), #cache_entry{key = Key, _ = '_'}) of
+        [] ->
+            gen_server:cast(Name, {increase_stat, miss}),
+            {error, not_found};
+        Matches ->
+            lists:foreach(fun(E) -> send_stat(Name, Now, E) end, Matches),
+            IsValid = fun(E) -> get_valid_value(Name, WaitForRefresh, Now, E) end,
```

Comment:
I think in order to keep consistent behavior and perform less list operations `IsValid` should be:
```
fun(E) -> 
    send_stat(Name, Now, E),
    get_valid_value(Name, WaitForRefresh, Now, E)
end,
```